### PR TITLE
Adding check for broken links in .json files 

### PR DESCRIPTION
**Pull request by xannem**
Our .json files in the https://github.com/finos/legend/tree/master/website/static/resource/studio/documentation sometimes have URL links inside them but they must be hard coded since other apps like https://github.com/finos/legend-studio makes usage of them. 

To check in case we have edited/modified files that the URLs depend on, we are adding a script that goes through each URL found in the .json file, tries to get their relative path, and throws an error if they cannot find it.

### CODE REVIEW

File: test-broken-links.js [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+  for (let i = 0; i < keys.length; i++) {
+    const key = keys[i];
+    if (jsonObject.entries[key].url == undefined) {
+      continue;
+    }
+    lookForUrl(jsonObject.entries[key].url);
+  }
+}
+
+async function lookForUrlWithAnchor(url) {
```
Comment (akphi) [inline test-broken-links.js:24]: very nice, I think the best way to do this though is to verify based on the static website content we build in the previous workflow step, that way, we are sure the paths work and the anchor are accounted for properly (with the current approach, anchors with dashes might pose a problem)

**Review by akphi — CHANGES_REQUESTED** [review]

File: test-broken-links.js (hunk 1 above)
Comment (xannem) [reply test-broken-links.js:24]: very true about the dashes 🙇

File: .github/workflows/docusaurus.yml [hunk 2]
```diff
@@ -47,11 +47,17 @@ jobs:
...
           ./build-site.sh
+      - name: Search for Broken Links
+        run: |
+          node ./test-broken-links.js
       - name: Publish Website
         run: |
           # Extract GitHub org/user
           REPO="${{ github.repository }}"
           GIT_USERNAME=${REPO%/*}
+          if [ "$GIT_USERNAME" != "finos" ]
```
Comment (akphi) [inline .github/workflows/docusaurus.yml:58]: don't need this. you can use if directly as part of the CI job: see https://docs.github.com/en/actions/using-jobs/using-conditions-to-control-job-execution

we can do something like
https://github.com/finos/legend-studio/blob/master/.github/workflows/release.yml#L17

**Review by akphi — CHANGES_REQUESTED** [review]
Please remove all files from `website/.docusaurus` and add them to `.gitignore`. Thanks!!

File: test-broken-links.js [hunk 3]
```diff
@@ -0,0 +1,72 @@
+const fs = require("fs");
```
Comment (akphi) [inline test-broken-links.js:1]: let's rename this file to `check-legend-application-documentation-broken-links.js`, also move it to `scripts/check-legend-application-documentation-broken-links.js` to start organize and not pollute the top-level of the repo.

File: .github/workflows/docusaurus.yml [hunk 4]
```diff
@@ -47,11 +47,17 @@ jobs:
       - name: Build Website
         run: |
           ./build-site.sh
```
Comment (akphi) [inline .github/workflows/docusaurus.yml:49]: help me move `build-site.sh` to `scripts/build-site.sh`, make sure you check if we use this script elsewhere.

File: test-broken-links.js [hunk 5]
```diff
@@ -0,0 +1,72 @@
...
+    locationOfFile +
+    url.substring(url.lastIndexOf("/"), url.lastIndexOf("#")) +
+    ".html";
+
+  const anchorTag = url.substring(url.lastIndexOf("#"));
+
+  if (await fs.existsSync(filePathName)) {
+    await fs.readFile(filePathName, "UTF-8", (err, data) => {
+      if (!data) {
+        throw new Error("Broken link found " + filePathName);
```
Comment (akphi) [inline test-broken-links.js:40]: I think we could improve this script slightly by collecting the errors and throw in the end if the list is non-empty. If you make errors here, I suppose you would have to run multiple times to find out.

File: test-broken-links.js [hunk 6]
```diff
@@ -0,0 +1,72 @@
...
+
+async function lookForUrl(url) {
+  currUrl = url.replace(WEBSITE_DIRECTORY, "");
+
+  lastDirectoryIndex = currUrl.lastIndexOf("/");
+
+  fileName = currUrl.substring(lastDirectoryIndex);
+  currUrlFinal = currUrl.substring(0, lastDirectoryIndex);
+
+  if (fs.existsSync("./website/build/" + currUrlFinal)) {
```
Comment (akphi) [inline test-broken-links.js:59]: let's create a constant for `const WEBSITE_CONTENT_DIRECTORY =  '/website/build';`

File: test-broken-links.js [hunk 7]
```diff
@@ -0,0 +1,72 @@
+const fs = require("fs");
+
+const WEBSITE_DIRECTORY = "https://legend.finos.org/";
+const DOC_DIRECTORY = "website/static/resource/studio/documentation/";
```
Comment (akphi) [inline test-broken-links.js:4]:
```suggestion
const APPLICATION_DOC_DIRECTORY = "website/static/resource/studio/documentation/";
```

File: test-broken-links.js [hunk 8]
```diff
@@ -0,0 +1,72 @@
+const fs = require("fs");
+
+const WEBSITE_DIRECTORY = "https://legend.finos.org/";
```
Comment (akphi) [inline test-broken-links.js:3]:
```suggestion
const DOC_WEBSITE_URL = "https://legend.finos.org/";
```

**Review by akphi — CHANGES_REQUESTED** [review]

File: scripts/check-legend-application-documentation-broken-links.js [hunk 9]
```diff
@@ -0,0 +1,81 @@
...
+
+  if (await fs.existsSync(filePathName)) {
+    await fs.readFile(filePathName, "UTF-8", (err, data) => {
+      if (!data) {
+        collectionOfErrors.push(["\n" + filePathName + "\n"]);
+      }
+
+      if (!data.toLocaleLowerCase().includes(anchorTag)) {
+        brokenUrls.push(["\n" + url + "\n"]);
+        throw new Error("Broken link(s) found: " + brokenUrls);
```
Comment (akphi) [inline scripts/check-legend-application-documentation-broken-links.js:52]: I don't think we need to throw here right?

```suggestion
```

File: scripts/check-legend-application-documentation-broken-links.js [hunk 10]
```diff
@@ -0,0 +1,81 @@
...
+  const anchorTag = url.substring(url.lastIndexOf("#"));
+
+  if (await fs.existsSync(filePathName)) {
+    await fs.readFile(filePathName, "UTF-8", (err, data) => {
+      if (!data) {
+        collectionOfErrors.push(["\n" + filePathName + "\n"]);
+      }
+
+      if (!data.toLocaleLowerCase().includes(anchorTag)) {
+        brokenUrls.push(["\n" + url + "\n"]);
```
Comment (akphi) [inline scripts/check-legend-application-documentation-broken-links.js:51]: why do we do this? you can just do `brokenUrls.push(url)` everywhere and when you print it out, you can print it nicely

File: scripts/check-legend-application-documentation-broken-links.js [hunk 11]
```diff
@@ -0,0 +1,81 @@
...
+    const key = keys[i];
+    if (jsonObject.entries[key].url == undefined) {
+      continue;
+    }
+    lookForUrl(jsonObject.entries[key].url);
+  }
+}
+
+if (brokenUrls.length !== 0) {
+  throw new Error("Broken link(s) found: " + brokenUrls);
```
Comment (akphi) [inline scripts/check-legend-application-documentation-broken-links.js:28]: Here, you can print it nicely, something like

```suggestion
  throw new Error(`Found ${brokenUrls.length} broken link(s):\n${brokenUrls.join('\n')}`);
```

**Review by akphi — CHANGES_REQUESTED** [review]

File: scripts/check-legend-application-documentation-broken-links.js [hunk 12]
```diff
@@ -0,0 +1,81 @@
+const fs = require("fs");
```
Comment (akphi) [inline scripts/check-legend-application-documentation-broken-links.js:1]: I generally agree with the logic of this file, but there are a few problems:
1. We're using fairly old JS syntax
2. The await/async logic seems odd and could be buggy so beware
3. The file name/file path resolution logic seems unnecessarily convoluted
4. We could improve support for external doc links: e.g. youtube, github wiki (with some anchors)

**Review by akphi — APPROVED** [review]

**Review by akphi — APPROVED** [review]

**Pull request merged**